File: src/Resources/boids.js
(() => {
    const JSB = typeof jsb !== 'undefined';

    const NATIVE_COMPUTATION = JSB && 1;
    const options = {
        boidCount: 8196,
        maxVelocity: 0.4,
        alignmentForce: 0.002,
        cohesionForce: 0.002,
        separationForce: 0.003,
        separationDistance: 0.1,
        flockmateRadius: 0.9,
        startingDelay: 2,
    };

    const { vec3, quat } = JSB ? glMatrix : require('./gl-matrix');
    const { root } = JSB ? chassis : require('./chassis');

    let tick = null;

    if (NATIVE_COMPUTATION) {

        jsb.BoidsManager.init(options);

        tick = (gTimeInMS) => {
            jsb.BoidsManager.tick(gTimeInMS);

            root.render();
        };

    } else {

        const parent = root.createTransform();
        const tempVec3a = vec3.create();
        const tempVec3b = vec3.create();
        const tempQuat = quat.create();
        const tempVec4 = quat.create();

        const HALF = vec3.set(vec3.create(), 0.5, 0.5, 0.5);

        const alignment = vec3.create();
        const cohesion = vec3.create();
        const separation = vec3.create();

        const getBoundaryFade = (v, clamp) => {
            return Math.min(clamp,
                1 - Math.abs(v[0]),
                1 - Math.abs(v[1]),
                1 - Math.abs(v[2])) / clamp;
        };

        const wrapBound = (v) => {
            if (v[0] > 1) { v[0] -= 2; }
            else if (v[0] < -1) { v[0] += 2; }
            if (v[1] > 1) { v[1] -= 2; }
            else if (v[1] < -1) { v[1] += 2; }
            if (v[2] > 1) { v[2] -= 2; }
            else if (v[2] < -1) { v[2] += 2; }
        };

        const clampLength = (v, max) => {
            const l = vec3.length(v);
            if (l > max) { vec3.scale(v, v, max / l); }
        };

        const applyForce = (acc, velocity, force, strength) => {
            vec3.scale(tempVec3a, force, options.maxVelocity / vec3.length(force));
            clampLength(vec3.subtract(tempVec3a, tempVec3a, velocity), strength);
            vec3.add(acc, acc, tempVec3a);
        };

        class Boid {
            model = root.createModel();
            transform = root.createTransform();

            acceleration = vec3.create();
            velocity = vec3.create();

            constructor() {
                this.transform.setParent(parent);
                this.model.setTransform(this.transform);

                vec3.random(tempVec3a);
                // tempVec3a[2] = Math.abs(tempVec3a[2]);
                this.transform.setPosition(tempVec3a);
                vec3.random(this.velocity, Math.random() * options.maxVelocity);

                this.update();
            }

            update() {
                vec3.normalize(tempVec4, this.velocity);
                quat.normalize(tempQuat, quat.set(tempQuat, tempVec4[2], 0, -tempVec4[0], 1 + tempVec4[1]));
                this.transform.setRotation(tempQuat);

                // vec3.normalize(tempVec4, this.acceleration); // visualize acceleration

                vec3.scaleAndAdd(tempVec4, HALF, tempVec4, 0.5);
                tempVec4[3] = getBoundaryFade(this.transform.getPosition(), 0.1);
                this.model.setColor(tempVec4);
            }
        }

        const boids = [];
        for (let i = 0; i < options.boidCount; ++i) {
            boids.push(new Boid());
        }

        let lastTime = -1;

        tick = (gTimeInMS) => {
            if (lastTime < options.startingDelay * 1000) {
                lastTime = gTimeInMS;
                root.render();
                return;
            }

            dt = (gTimeInMS - lastTime) / 1000;
            lastTime = gTimeInMS;
            let distance = 0;

            for (const b1 of boids) {
                vec3.set(alignment, 0, 0, 0);
                vec3.set(cohesion, 0, 0, 0);
                vec3.set(separation, 0, 0, 0);
                let alignmentActive = false;
                let cohesionActive = false;
                let separationActive = false;

                for (const b2 of boids) {
                    if (b1 === b2) { continue; }
                    vec3.subtract(tempVec3a, b2.transform.getPosition(), b1.transform.getPosition());
                    distance = Math.max(0.01, vec3.length(tempVec3a) - 0.1);

                    if (distance < options.separationDistance) {
                        vec3.scale(tempVec3b, tempVec3a, -1 / distance);
                        vec3.add(separation, separation, tempVec3b);
                        separationActive = true;
                    }

                    if (distance < options.flockmateRadius) {
                        vec3.add(cohesion, cohesion, tempVec3a);
                        cohesionActive = true;
                        vec3.add(alignment, alignment, b2.velocity);
                        alignmentActive = true;
                    }
                }

                const { acceleration: acc, velocity: vel } = b1;
                vec3.set(acc, 0, 0, 0);
                if (alignmentActive) { applyForce(acc, vel, alignment, options.alignmentForce); }
                if (cohesionActive) { applyForce(acc, vel, cohesion, options.cohesionForce); }
                if (separationActive) { applyForce(acc, vel, separation, options.separationForce); }
            }

            for (const b of boids) {
                clampLength(vec3.add(b.velocity, b.velocity, b.acceleration), options.maxVelocity);
                wrapBound(vec3.scaleAndAdd(tempVec3a, b.transform.getPosition(), b.velocity, dt));
                b.transform.setPosition(tempVec3a);
                b.update();
            }

            root.render();
        };
    }

    if (JSB) { window.gameTick = tick; }
    else { tick(); console.log('tick finished.'); }

})();

File: tests/ScriptTest/Boids.cpp
#include "Boids.h"

#include <vector>
#include "Chassis.h"
#include "base/Log.h"

using namespace cc::vmath; // NOLINT(google-build-using-namespace) when inside cpp files this is usually fine

namespace application {

template <typename Value>
struct Boid {
    using Index = cc::vmath::replace_scalar_t<Value, cc::vmath::Index>;
    using Vec3  = cc::vmath::Vec3<Value>;

    Index model{-1};
    Index transform{-1};

    Vec3 acceleration = cc::vmath::zero<Vec3>();
    Vec3 velocity     = cc::vmath::zero<Vec3>();

    // cache states
    Vec3 position = cc::vmath::zero<Vec3>();

    // NOLINTNEXTLINE(google-explicit-constructor) false positive when involving __VA_ARGS__
    CC_VMATH_STRUCT(Boid, model, transform, acceleration, velocity, position)
};

CC_VMATH_STRUCT_SUPPORT_1(application, Boid, model, transform, acceleration, velocity, position)

using BoidF = Boid<float>;
using BoidP = Boid<cc::vmath::FloatP>;
using BoidX = Boid<cc::vmath::FloatX>;

namespace {
std::vector<cc::TransformView *> transformViews;
std::vector<cc::ModelView *>     modelViews;
BoidsOptions                     options;
BoidX                            boids;
} // namespace

template <typename Value>
static Vec3<Value> sampleUniformSphere(PCG32<Value> &rng) {
    Value theta  = rng.next_float32() * cc::math::PI * 2.F;
    Value cosPhi = rng.next_float32() * 2.F - 1.F;
    Value sinPhi = sqrt(1.F - sqr(cosPhi));
    auto [s, c]  = sincos(theta);
    return Vec3<Value>{c * sinPhi, s * sinPhi, cosPhi};
}

template <typename Value_, typename Value = expr_t<Value_>>
static float getBoundaryFade(Vec3<Value_> v, float clamp) {
    static const Vec3<Value> OFFSET{0.F, 0.F, -0.5F};
    static const Vec3<Value> SCALE{1.F, 1.F, 2.F};

    return hmin(concat(1.F - abs(v + OFFSET) * SCALE, clamp)) / clamp;
}

template <typename Value_, typename Value = expr_t<Value_>>
static void update(const Boid<Value_> &boid) {
    auto dir = normalize(boid.velocity);
    auto rot = normalize(Vec4<Value>{dir.z(), 0.F, -dir.x(), 1.F + dir.y()});

    // visualize acceleration
    // dir = normalize(boid.acceleration);

    dir = fmadd(dir, 0.5F, 0.5F);
    for (size_t i = 0U; i < Value::Size; ++i) {
        if (!boid.transform[i]) break; // after the last one
        auto * transformView = transformViews[boid.transform[i]];
        auto * modelView     = modelViews[boid.model[i]];
        auto &&pos           = slice(boid.position, i);
        transformView->setRotation(slice(rot, i));
        transformView->setPosition(pos);
        modelView->setColor(concat(slice(dir, i), getBoundaryFade(pos, 0.1F)));
    }
}

template <typename Value_, typename Value = expr_t<Value_>>
static Vec3<Value> &&wrapBound(Vec3<Value_> &&v) {
    v.x()[v.x() > 1.F] -= 2.F;
    v.x()[v.x() < -1.F] += 2.F;
    v.y()[v.y() > 1.F] -= 2.F;
    v.y()[v.y() < -1.F] += 2.F;
    v.z()[v.z() > 1.F] -= 1.F;
    v.z()[v.z() < 0.F] += 1.F;
    return std::move(v);
}

template <typename Value_, typename Value = expr_t<Value_>>
static Vec3<Value> clampLength(const Vec3<Value_> &v, float max) {
    auto l = dot(v, v);
    return select(l > sqr(max), v * max * rsqrt(l), v);
}

template <typename Value_, typename Value = expr_t<Value_>>
static Vec3<Value> applyForce(const Vec3<Value_> &velocity, const Vec3<Value> &force, float strength) {
    return clampLength(fmsub(normalize(force), options.maxVelocity, velocity), strength);
}

void BoidsManager::init(const BoidsOptions &newOptions) {
    options = newOptions;

    cc::Root *         root   = cc::RootManager::getInstance();
    cc::TransformView *parent = root->createTransform();
    transformViews.push_back(parent);

    setSlices(boids, options.boidCount);
    for (uint32_t i = 0; i < options.boidCount; ++i) {
        cc::TransformView *transform = root->createTransform();
        cc::ModelView *    model     = root->createModel();

        transform->setParent(parent);
        model->setTransform(transform);

        slice(boids.model, i)     = static_cast<Index>(modelViews.size());
        slice(boids.transform, i) = static_cast<Index>(transformViews.size());

        transformViews.push_back(transform);
        modelViews.push_back(model);
    }

    CC_LOG_INFO("Boids count: %d", options.boidCount);

    PCG32<FloatP> rng(std::time(nullptr)); // make every run unique
    for (size_t i = 0; i < packets(boids); ++i) {
        auto &&p       = packet(boids, i);
        p.position     = sampleUniformSphere(rng);
        p.position.z() = abs(p.position.z());
        p.velocity     = sampleUniformSphere(rng) * options.maxVelocity;
        update(p);
    }
}

void BoidsManager::destroy() {
    for (auto *transform : transformViews) {
        CC_SAFE_DELETE(transform);
    }
    transformViews.clear();

    for (auto *model : modelViews) {
        CC_SAFE_DELETE(model);
    }
    modelViews.clear();
}

void BoidsManager::tick(float globalTime) {
    static float lastTime = -1.F;
    if (lastTime < options.startingDelay * 1000.F) {
        lastTime = globalTime;
        return;
    }

    float dt = (globalTime - lastTime) / 1000.F;
    lastTime = globalTime;

    static const auto ZERO = zero<Vec3P>();

    static Vec3P alignment;
    static Vec3P cohesion;
    static Vec3P separation;
    static MaskP alignmentActive;
    static MaskP cohesionActive;
    static MaskP separationActive;

    for (size_t i = 0; i < packets(boids); ++i) {
        auto &&b1        = packet(boids, i);
        alignment        = zero<Vec3P>();
        cohesion         = zero<Vec3P>();
        separation       = zero<Vec3P>();
        alignmentActive  = false;
        cohesionActive   = false;
        separationActive = false;

        for (size_t j = 0; j < slices(boids); ++j) {
            auto &&b2        = slice(boids, j);
            Vec3P  dir       = b2.position - b1.position;
            FloatP lengthSq  = dot(dir, dir);
            FloatP distance  = max(0.01F, sqrt(lengthSq) - 0.1F);
            MaskP  allOthers = lengthSq > 0.F;

            MaskP shouldSeparate = allOthers & (distance < options.separationDistance);
            separation += select(shouldSeparate, dir * rcp(-distance), ZERO);
            separationActive |= shouldSeparate;

            MaskP shouldConnect = allOthers & (distance < options.flockmateRadius);
            cohesion += select(shouldConnect, dir, ZERO);
            cohesionActive |= shouldConnect;
            alignment += select(shouldConnect, b2.velocity, ZERO);
            alignmentActive |= shouldConnect;
        }

        b1.acceleration = select(alignmentActive, applyForce(b1.velocity, alignment, options.alignmentForce), ZERO);
        b1.acceleration += select(cohesionActive, applyForce(b1.velocity, cohesion, options.cohesionForce), ZERO);
        b1.acceleration += select(separationActive, applyForce(b1.velocity, separation, options.separationForce), ZERO);
    }

    for (size_t i = 0; i < packets(boids); ++i) {
        auto &&b   = packet(boids, i);
        b.velocity = clampLength(b.velocity + b.acceleration, options.maxVelocity);
        b.position = wrapBound(b.position + b.velocity * dt);
        update(b);
    }
}

} // namespace application
